### PR TITLE
Add support for local config

### LINKED ISSUE #69: cookieValidationKey is not generated after installation
### What steps will reproduce the problem?

```
composer install
```
### What's expected?

Security code is generated and inserted in `web.php`
### What do you get instead?

I had to put it manually.
### Additional info

PHP version: 7.0.4
Composer version 1.2-dev (08ef916beded8c738be11acc4f463c450e277d64) 2016-06-06 10:22:06
#### Php modules

```
[PHP Modules]
bcmath
calendar
Core
ctype
curl
date
dom
exif
fileinfo
filter
ftp
gd
gettext
hash
iconv
json
libxml
mbstring
mcrypt
mongodb
openssl
pcntl
pcre
PDO
pdo_sqlite
Phar
posix
readline
Reflection
session
shmop
SimpleXML
soap
sockets
SPL
sqlite3
standard
sysvmsg
sysvsem
sysvshm
tokenizer
wddx
xdebug
xml
xmlreader
xmlwriter
xsl
Zend OPcache
zlib

[Zend Modules]
Xdebug
Zend OPcache
```

### PR DESCRIPTION
**Pull request by rob006**
| Q | A |
| --- | --- |
| Is bugfix? | no |
| New feature? | yes |
| Breaks BC? | no |
| Tests pass? | yes |
| Fixed issues | #69 |

replaces #60
closes #69

Simplified version of #60 without environment-specific config files.

### CODE REVIEW

**Comment by samdark** [comment]
@yiisoft/core-developers need opinions about these changes.

File: composer.json [hunk 1]
```diff
@@ -29,6 +29,12 @@
         "process-timeout": 1800
     },
     "scripts": {
+         "post-install-cmd": [
+            "php -r \"if(!file_exists('config/console-local.php')){copy('config/templates/console-local.php', 'config/console-local.php');}\"",
```
Comment (SilverFire) [inline composer.json]: I'd prefer moving these operations to the separate method of `yii\composer\Installer` class.

Comment (rob006) [inline composer.json]: https://github.com/yiisoft/yii2-app-basic/pull/60#discussion_r65929692

**Comment by SilverFire** [comment]
The reasons for creating local configs are not obvious for beginners. The idea should be well documented: why local config is under `.gitignore`, when should I use local config and when should I the common one

**Comment by samdark** [comment]
Yeah. It's not obvious for advanced app users as well despite being documented.

**Comment by rob006** [comment]
Synced with master and added docs about local configs. Gladly would see some progress for my PRs for `yii2-composer` to push this forward.

**Comment by thyseus** [comment]
+1

**Comment by cebe** [comment]
I am for adding a simple local config concept to the basic application because it is a needed feature in nearly every project. But it should not make things too complex.
So I'd like to propose the following:

local configs should not use `ArrayHelper::merge` as it can produce weird situations sometimes when merging sub arrays. I'd prefer a config file included like this:

main.php:

```php
<?php

$config = [
    // ...
];

if (file_exists($localConfig = __DIR__ . '/main.local.php')) {
    require($localConfig);
}
return $config;
```

main.local.php:

```php
<?php

$config['components']['redis'] = 'yii\redis\Connection';

```

Array manipulation is more basic and easier to understand than recursive merge.

A minor thing about naming, I prefer `main.local.php` instead of `main-local.php` as I find it more intuitive that main belongs to main.local but I have no strong opinon on that one.

**Comment by samdark** [comment]
Agree.

**Comment by rob006** [comment]
> local configs should not use ArrayHelper::merge as it can produce weird situations sometimes when merging sub arrays.

About what situations you're talking about?

**Comment by cebe** [comment]
example:
main.php:
```php
[
    'params' => [
        'adminUsers' => ['cebe', 'samdark'],
    ],
]
```

main.local.php:
```php
[
    'params' => [
        'adminUsers' => ['rob006'],
    ],
]
```

the final result will contain all users. that is not expected.

**Comment by rob006** [comment]
I would say that this is expected - this is how `ArrayHelper::merge()` works, and it affects not only configuration. If developer does not understand this, he probably breaks something sooner or later. 

My approach also scales much better - if you have many custom local settings, storing them as array is much more readable. If you decide to extract settings shared between console and web app (to `main.php` config file), you just need to add one line into `index.php` and `yii` files. Same for adding environment-specific config files or `main-local.php` (local config shared between console and web app). With @cebe's approach you need to reinvent wheel again and probably end up with my solution. 

About `-local.php` suffix - this is convention taken from advanced app template. I think it is better to keep this consistent.

**Comment by cebe** [comment]
> I would say that this is expected - this is how ArrayHelper::merge() works, and it affects not only configuration. If developer does not understand this, he probably breaks something sooner or later.

The main point about not introducing `ArrayHelper` here is that config is one of the first thing a new user has to deal with to get started, e.g. set up the database. If there are too many files around it is likely to get confused about which files are merged in which way. Configuration should be straight forward in basic app. I agree that `ArrayHelper::merge()` can be useful in many ways when you have enough experience but it introduces too much complexity for beginners imo.

I feel the need for a third application template between basic and advanced here.

**Comment by dynasource** [comment]
> If there are too many files around it is likely to get confused about which files are merged in which way

the less files the better. IMO its better to start as small as possible and give advice how to expand than the other way around. Alls these config files are not making the life of a developer easier (especially when you don't have experience with it). 

In all of my projects I am now trying to minimize the number of config files. The less you have, the less you have to manage.

**Comment by cebe** [comment]
How about adding the code that will include a local config file if it exists but do not create the templates?

**Comment by dynasource** [comment]
looks good. I like it better than the array merging because its more tightly connected to its owner.

**Comment by rob006** [comment]
We need at least two templates for local configs: one for DB settings and second for `cookieValidationKey` in web application config.

**Comment by samdark** [comment]
Something like https://github.com/samdark/yiiframework-ru/blob/master/util/Configurator.php?

**Comment by rob006** [comment]
Sorry guys but implementing changes proposed by @cebe require more work than I suspected, and I don't want to burn more time on this since I'm not using this template anymore. Consider this PR as finished. If you want additional changes on this, someone else has to do it.

**Comment by cebe** [comment]
@rob006 no problem, thank you for your contribution! We will decide on how to proceed with this later and make the necessary changes.

**Comment by samdark** [comment]
OK. Thank you for contributing!

**Comment by SilverFire** [comment]
In result of internal discussion, we decided that this PR brings too much complexity as for **basic** application template so the PR will be closed.

@rob006 thank you for your contribution and sorry that this one is rejected.

**Pull request closed without merging**